### ISSUE TITLE
Cut instead of copy

### ISSUE THREAD

**Issue by raffaem** [sqlitebrowser/sqlitebrowser#2775]
Details for the issue
--------------------

#### What did you do?

1. Open a sqlite database
2. Swtich to "Execute SQL" tab
3. Select some SQL code and CTRL+C
4. Switch to another window ***while keeping the code highlighted***

#### What did you expect to see?

The highlighted code should stay there

#### What did you see instead?

The highlighted code is cut and removed

Useful extra information
-------------------------

The info below often helps, please fill it out if you're able to. :)

#### What operating system are you using?

- [ ] Windows: ( _version:_ ___ )
- [X] Linux: ( _distro:_ Fedora Workstation 34 )
- [ ] macOS: ( _version:_ ___ )
- [ ] Other: ___

#### What is your DB4S version?

- [X] 3.12.2
- [ ] 3.12.1
- [ ] 3.12.0
- [ ] 3.11.x
- [ ] Other: ___

#### Did you also

- [ ] Try out the latest nightly version: https://github.com/sqlitebrowser/sqlitebrowser#nightly-builds
- [X] Search for an existing similar issue: https://github.com/sqlitebrowser/sqlitebrowser/issues?utf8=%E2%9C%93&q=is%3Aissue%20

**Comment by scottfurry** [comment]
Using latest from repository (3.12.99) on Gentoo (5.10.49).
I **cannot** confirm this bug.

As 'V', 'C', and 'X' (paste, copy, and cut respectively) are close to each other on typical QWERTY keyboard, I suspect other issues are causing this problem.

Would recommend ensuring keyboard mechanics are sound (keys are clean and functional when individually used). If this is a laptop or small keyboard, there is also the possibility of excessive pressure on key press warping keyboard causing adjacent key press to be triggered. Again, keyboard mechanics.

Last resort would be to double check locale settings. Certain locales or regional settings can change selected keyboard layout.

In short, I don't think it's code. Other things could and should be checked out before saying software is bad.

**Comment by chrisjlocke** [comment]
> Switch to another window while keeping the code highlighted

So at what point is the text removed?  You highlight the text and press CTRL+C.  The text is there and highlighted,  You switch to another application.  You switch back to DB4S.  The text is now gone?
Is the text gone every time?
If you switch apps and press CTRL+V then switch back to DB4S is the text gone?  Has the text gone (with nothing there) or is the text replaced with a space or something?

**Comment by raffaem** [comment]
> > Switch to another window while keeping the code highlighted
> 
> So at what point is the text removed? You highlight the text and press CTRL+C. The text is there and highlighted, You switch to another application. You switch back to DB4S. The text is now gone?
> Is the text gone every time?
> If you switch apps and press CTRL+V then switch back to DB4S is the text gone? Has the text gone (with nothing there) or is the text replaced with a space or something?

I made a video that shows the bug [here](https://youtu.be/6ZHG4WagQKQ)

In the video I press CTRL+C at around 0:05. Then at 0:06 I switch windows using ALT+TAB.

I found out it only happens when there is a database opened.

**Comment by chrisjlocke** [comment]
Thanks for the video, and for clarifying it depends on what application is selected.  This would indicate it couldn't be DB4S code per se - otherwise it would happen regardless of the application.  Sounds more like some weird keyboard combination is being sent/received and misinterpreted somehow.  The cutting / removing of text is within the code of the widget used, and not within DB4S itself.  Could be some weird Qt bug linked to the OS.  Would be interesting to know if it occurs with another Qt-based app, but finding another on the same Qt version using the same widget would be extremely difficult.

**Comment by scottfurry** [comment]
This leaves me the impression that there is some mis-configuration of keyboard shortcuts that may be conflated with a problem of keyboard layouts and locales.

For example, if you have set your language but keyboard is for a European/International layout programs would be confused as you have described above. I have run into this where pressing a key doesn't produce on the screen what is printed on the keyboard key.

Also, confused locales would cause differences in how different programs behave. Some programs will read different aspects of locale settings. (LANG, LC_CTYPE or LC_ALL). Inconsistencies in locale settings would cause various programs to produce different behaviours.

My recommendation is to ping the fedora forms/documents for the following:
- using a single locale, language setting (e.g. en_US, fr_FR, de_DE et al)
- ensure you have the proper keyboard layout for your model of keyboard

Fix this, with reboots after changes, and try again.

**Comment by raffaem** [comment]
@scottfurry I'll take your suggestion and ping the Fedora forums about it.

Honestly I have copied and pasted from and into a lot of applications and sqlitebrowser it's the only application where this behavior happens.

Let's keep this open for a while until I get something from the Fedora forum.

**Comment by scottfurry** [comment]
@raffaem see [this](https://docs.fedoraproject.org/en-US/fedora/rawhide/system-administrators-guide/basic-system-configuration/System_Locale_and_Keyboard_Configuration/) document.
Paste the output of the command `locale status`

If it is only happening in DB4S, then we may have a locale problem in the program not being properly addressed.

**Comment by raffaem** [comment]
```
$ localectl status
   System Locale: LANG=en_US.UTF-8
                  LC_NUMERIC=en_GB.UTF-8
                  LC_TIME=en_GB.UTF-8
                  LC_MONETARY=en_GB.UTF-8
                  LC_PAPER=en_GB.UTF-8
                  LC_MEASUREMENT=en_GB.UTF-8
       VC Keymap: it
      X11 Layout: it
```

**Comment by scottfurry** [comment]
@raffaem  And that speaks volumes to me. Language is set to English(US) with English(UK) variants for money, time, measurements, etc(metric w/ pound symbol and A4 paper et al).
**However**, why the Italian Keyboard is beyond me. Square those away to a single set appropriate for you and we'll try the cut, copy, past again.

**Comment by raffaem** [comment]
> However, why the Italian Keyboard is beyond me. Square those away to a single set appropriate for you and we'll try the cut, copy, past again.

My keyboard has the Italian layout. I bought it in Italy and it has the Italian layout. With the Italian layout, all the characters written on the screen corresponds to what is printed on the keyboard. Every single time for all of them. With any other layout, the characters that gets written on the screen don't correspond to what is written on the keyboard.

However I prefer the English language because it's easy to find out instructions and report bug, and the UK measurement system because I don't like the US mm/dd/yyyy way of spelling out the dates.

I will try to set everything to Italian and try out again.

**Comment by raffaem** [comment]
So with these settings:

```
$ localectl status
   System Locale: LANG=it_IT.UTF-8
       VC Keymap: it
      X11 Layout: it
```

The bug still happens.

Moreover the text gets erased also if I switch to Firefox. It gets erased if I switch to the Terminal. But it doesn't get erased if I switch to Thunderbird.

I guess it depends on whether the application you switch to "expects" something to be pasted into.

**Comment by scottfurry** [comment]
> However I prefer the English language because it's easier to find out documentation and instructions, and the UK measurement system because I don't like the US mm/dd/yyyy way of spelling out the dates.

You and me both - https://xkcd.com/1179/

I think the confusion between the defined system language (LANG = en_US.UTF-8) and the Italian keyboard layout may be causing some grief. **Generally**, regional settings are handled by the underlying QT library. Other than passing text into translation, DB4S doesn't do anything special WRT regional settings beyond that.

I would look at some kind of keyboard capture. IIRC, there is something in X or gnome that shows the decoded keyboard presses. My suspicion is that the keyboard codes are being misinterpreted resulting in problems you are observing. However, mechanical issues are not wholly off the table, either.

**Comment by raffaem** [comment]
If I set everything to English (US), including the keyboard layout

![image](https://user-images.githubusercontent.com/54762742/126460244-336fc0b5-0d13-4bf8-9d87-c506984c2a60.png)

The localectl still displays  `it` as the keyboard layout:

```
$ localectl status
   System Locale: LANG=en_US.UTF-8
       VC Keymap: it
      X11 Layout: it
```

And the bug still happens.

The problem with the mechanical issues of my keyboard is that it happens when I switch windows. If CTRL+X got passed instead of CTRL+C the text would get removed right away

**Comment by scottfurry** [comment]
It's the removal of text when switching applications that bothers me the most. CTRL+C just copies. CTRL+X is cut and is,IIRC, mostly application dependent. For example in a file manager, selecting a file and pressing CTRL+X doesn't do anything to the file until pasting, or  CTRL+V, in file manager. In editors, cutting takes effect immediately by removing selected text.

**Comment by mgrojo** [comment]
We have to find out what version of QScintilla is used by your
sqlitebrowser build. It seems a bug in that part. Maybe they are compiling
with their own version of it?

I guess the problem is with the loss of focus and the impact on selection.
The new application might be claiming ownership of the X11 primary
selection, but doesn't explain per se why our selection is deleted.

**Comment by raffaem** [comment]
If we fix #2774 I can try the AppImage version instead of the Fedora provided one

**Comment by mgrojo** [comment]
I see in  #2611 that the Fedora version is depending on a Fedora's qscintilla package. That package might be the source of the problem. This might of interest to @manisandro

**Comment by camtofu** [comment]
I can confirm this bug (op mentioned).
Platform: Debian 10.10 x64, Gnome 3.20
DB4S 3.12.2 for Linux, both source code and Appimage build.
Affected UI elements are "Execute SQL" and "Edit Database Cell" in DB4S.
"Browser Data" are not affected.
When swith back DB4S from backgroiund to foreground, formerly selected text in a focused  "Execute SQL" or "Edit Database Cell" is deleted by modifier key (ALT, CTRL, SHIFT) press.
This bug seems related to qscintilla widget.
But in DB4S 3.10.1 for Linux (provided by Debian 10 official apt repository) is not affected by this bug.

**Comment by chrisjlocke** [comment]
Eurk.  Thanks for taking the time to report your findings. 👍

**Comment by uvtofu** [comment]
I confirm this bug still exist in latest linux appimage build (2022-05-20 download).
